File: blockchain-parser/src/redis.js
const Redis = require('ioredis');
const config = require('./config');

class RedisService {
  constructor() {
    this.redis = new Redis(config.redis);
  }

  async getLastProcessedBlock() {
    return await this.redis.get('last_processed_block');
  }

  async setLastProcessedBlock(blockNumber) {
    return await this.redis.set('last_processed_block', blockNumber);
  }

  async isTxProcessing(txHash) {
    return await this.redis.get(`tx:processing:${txHash}`);
  }

  async setTxProcessing(txHash, value = '1', expire = 300) {
    return await this.redis.set(`tx:processing:${txHash}`, value, 'EX', expire);
  }

  async removeTxProcessing(txHash) {
    return await this.redis.del(`tx:processing:${txHash}`);
  }

  async cacheUserAddress(address, userId) {
    return await this.redis.set(`address:user:${address}`, userId, 'EX', 3600);
  }

  async getCachedUserByAddress(address) {
    return await this.redis.get(`address:user:${address}`);
  }
}

module.exports = new RedisService();

File: blockchain-parser-go/redis/redis.go
package redis

import (
	"context"
	"strconv"
	"time"

	"github.com/go-redis/redis/v8"
)

type RedisClient struct {
	client *redis.Client
	ctx    context.Context
}

func NewRedisClient(host string, port int, password string) (*RedisClient, error) {
	client := redis.NewClient(&redis.Options{
		Addr:     host + ":" + strconv.Itoa(port),
		Password: password,
		DB:       0,
	})

	ctx := context.Background()
	if err := client.Ping(ctx).Err(); err != nil {
		return nil, err
	}

	return &RedisClient{
		client: client,
		ctx:    ctx,
	}, nil
}

func (r *RedisClient) GetLastProcessedBlock() (int64, error) {
	val, err := r.client.Get(r.ctx, "last_processed_block").Int64()
	if err == redis.Nil {
		return 0, nil
	}
	return val, err
}

func (r *RedisClient) SetLastProcessedBlock(blockNumber int64) error {
	return r.client.Set(r.ctx, "last_processed_block", blockNumber, 0).Err()
}

func (r *RedisClient) IsTxProcessing(txHash string) (bool, error) {
	val, err := r.client.Exists(r.ctx, "tx:processing:"+txHash).Result()
	return val > 0, err
}

func (r *RedisClient) SetTxProcessing(txHash string, expire time.Duration) error {
	return r.client.Set(r.ctx, "tx:processing:"+txHash, "1", expire).Err()
}

func (r *RedisClient) RemoveTxProcessing(txHash string) error {
	return r.client.Del(r.ctx, "tx:processing:"+txHash).Err()
}

func (r *RedisClient) CacheUserAddress(address string, userID int) error {
	return r.client.Set(r.ctx, "address:user:"+address, userID, time.Hour).Err()
}

func (r *RedisClient) GetCachedUserByAddress(address string) (int, error) {
	val, err := r.client.Get(r.ctx, "address:user:"+address).Int()
	if err == redis.Nil {
		return 0, nil
	}
	return val, err
}
